File: LLD/design patterns/builder.py
class HttpRequest():
    def __init__(self, url, method, headers, param=None, body=None, timeout=None):
        self.url=url
        self.method=method
        self.header=headers
        # self.param=param
        # self.body=body
        # self.timeout=timeout
        print(f"{self.url=} {self.method=} {self.header=}")


httpRequest=HttpRequest("https://api.example.com/data", method="POST", headers="header")


class HttpRequestBuilder():

    class Builder():
        def __init__(self):
            self.url = None
            self.method = None
            self.header = None
        def addUrl(self, url):
            self.url=url
            return self
        def addMethod(self, method):
            self.method=method
            return self
        def addHeader(self, header):
            self.header=header
            return self
        def build(self):
            return HttpRequestBuilder(self)

    def __init__(self, builder:Builder):
        self.url=builder.url
        self.method=builder.method
        self.header=builder.header
        print(f"{self.url=} {self.method=} {self.header=}")

httpRequestBuilder=HttpRequestBuilder.Builder()\
                       .addUrl("https://api.example.com/data")\
                        .addMethod("POST")\
                        .addHeader("header").build()

File: cpp/lld_cpp/builder.cpp
//
// Created by 2025 on 4/8/2026.
//
#include "../header.h"
class HttpRequest {

public:
    string url;
    string method;
    map<string,string> header;
    string params;
    string body;
    int timeout;


    class Builder {
        string url;
        string method="GET";
        string header;
        string  params;
        string body;
        int timeout=3000;
    public:
        Builder(string &url): url(url){}
        Builder& setMethod(string &method) {
            this->method=method;
            return *this;
        }
        Builder& addHeader(string &header) {
            this->header=header;
            return *this;
        }
        Builder& addParam(string &param) {
            this->params=param;
            return *this;
        }
    };
};

int main(int argc, char *argv[]) {

}
